csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,252 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+  int n_outer_loop = [&sm_count, &reduction_tvs, &runtime_info]() -> int {
+    for (auto buffer : reduction_tvs) {
+      if (scheduler_utils::isFastestDimReduction(buffer)) {
+        continue;
+      }
+      int64_t outer_dim_size = -1;
+      for (auto id : buffer->getMaybeRFactorDomain()) {
+        if (id->isReduction()) {
+          auto id_size =
+              runtime_info.expressionEvaluator().evaluate(id->extent());
+          NVF_ERROR(id_size.hasValue(), "Could not infer outer dim size.");
+          if (outer_dim_size == -1) {
+            outer_dim_size = id_size.as<int64_t>();
+          } else {
+            outer_dim_size *= id_size.as<int64_t>();
+          }
+        }
+      }
+      int gdimy = sm_count;
+      int n_loops = (int)ceilDiv(outer_dim_size, gdimy);
+      return n_loops;
+    }
+    NVF_ERROR(false, "Could not infer outer loop count.");
+    return 1;
+  }();
+
+  // set latency for each buffer.
+  // Assume load from smem to register has a latency of 1, from gmem to smem has
+  // a latency of 10. Total latency is proportional to data type size. Reused in
+  // each iteration if the buffer is an outer broadcast tensor.
+  constexpr int gmem_to_smem_latency = 10;
+  for (auto idx = 0; idx < n_buffer_inputs; idx++) {
+    int latency = 0;
+    auto input_buffer = projectable_buffer_inputs.at(idx);
+    bool is_reused = std::any_of(
+        outer_broadcast_tvs.begin(),
+        outer_broadcast_tvs.end(),
+        [&input_buffer](TensorView* tv) {
+          return DependencyCheck::isDependencyOf(input_buffer, tv);
+        });
+    // gmem to smem
+    latency += gmem_to_smem_latency * (is_reused ? 1 : n_outer_loop);
+    // smem to register
+    latency += n_outer_loop * n_proj_buffer_uses.at(idx);
+    int dtype_size = (int)dataTypeSize(input_buffer->getDataType().value());
+    buffer_latency[idx] = latency * dtype_size;
+  }
+
+  // reorder projectable_buffer_inputs based on the latency from smallest to
+  // largest
+  std::vector<int> idxs(n_buffer_inputs);
+  std::iota(idxs.begin(), idxs.end(), 0);
+  std::stable_sort(idxs.begin(), idxs.end(), [&buffer_latency](int i, int j) {
+    return buffer_latency[i] < buffer_latency[j];
+  });
+  std::vector<TensorView*> sorted_candidate_tvs;
+  sorted_candidate_tvs.reserve(n_buffer_inputs);
+  for (auto idx : idxs) {
+    sorted_candidate_tvs.emplace_back(projectable_buffer_inputs.at(idx));
+  }
+  return sorted_candidate_tvs;
+}
+
+// Sort persistent buffers based on the number of uses from smallest to largest.
+// Used when buffers are not projected to inputs.
+std::vector<TensorView*> sortPersistentBuffers(
+    std::vector<TensorView*> persistent_buffers) {
+  // Directly sort the persistent_buffers vector based on the number of uses.
+  std::sort(
+      persistent_buffers.begin(),
+      persistent_buffers.end(),
+      [](TensorView* a, TensorView* b) {
+        return a->uses().size() < b->uses().size();
```

Comment:
The number of uses only reflects the static count and is not necessarily correlated with the actual use count. I suppose you found this static info is enough for the layernorm case?

Reply:
Added comment
```
// TODO: Currently, this function is unused as all existing cases (LayerNorm bwd
// and RMS norm bwd) involve projecting buffers to inputs. Consider this
// function a placeholder for scenarios where buffers are not projected to
// inputs. Refer to the test CombinedSchedulerInnerOuterNoOuterBroadcastTv for a
// concrete example of such cases. Add NoOuterBroadcast cases to our benchmarks
// and test the performance of this function.
```

Reply:
Sorry, I'm not sure how to interpret your comment. Is this enough or not?

Reply:
It is enough for layernorm case.